### PR TITLE
[RFC] design doc: proposal for off-cpu profiling

### PR DESCRIPTION
**Pull request by florianl**
This proposal builds on and follows the approach described in https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/114.

### CODE REVIEW

**Review by felixge — APPROVED** [review]
This is a very promising proposal, thanks for submitting it 🙇 .

I think only the capabilities in Option B makes sense to me. Capturing Off-CPU stack trace samples without duration doesn't see useful to me.

The other important question is sampling. I suspect that sampling based on Off-CPU duration is going to produce more useful results than giving each event an equal chance of being captured. JFR implements this with thresholds, but we could also explore something along the lines of Go's Allocation sampling strategy that sets randomized sampling points drawn from an exponential distribution with a desired mean.

As far as other implementation details are concerned, it might be worth considering the following prior art:

* [Off-CPU Analysis](https://www.brendangregg.com/offcpuanalysis.html) article from Brendan Gregg
* [bcc's offcputime.py](https://github.com/iovisor/bcc/blob/master/tools/offcputime.py): Which seems to only use a `finish_task_switch` kprobe without instrumenting a tracepoint.
* [Perfetto's CPU Scheduling Event Docs](https://perfetto.dev/docs/data-sources/cpu-scheduling): Which discusses using `sched_waking` as an alternative to `sched_switch`.

My approval indicates that I support the general proposal along the lines of option B, and that we should to hash out sampling strategies and implementation details during prototyping.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:92]: Yes - the assumption is, that the introduced overhead is largely from unwinding the stack. To unwind the stack, multiple eBPF helpers are called, e.g. to read memory, read values from eBPF maps or update eBPF mpas. Unfortunately I can't provide number to back this assumption.

Comment (felixge) [reply design-docs/00001-off-cpu-profiling/README.md:92]: Thanks for clarifying. Feel free to resolve this thread. This can be confirmed in prototyping.

**Comment by tsint** [comment]
Considering that off-CPU profiling is more expensive, it is recommended to add a process filter so that only processes meeting certain criteria are analyzed.

**Comment by florianl** [comment]
> [..], it is recommended to add a process filter so that only processes meeting certain criteria are analyzed.

Filtering for a specific process is, from my point of view, not a viable option, as
- the complete architecture of the profiling agent is focused on complete system profiling, so the architecture would need to change.
- filtering for a process would not mitigate the named risks
- filtering for a process would also add significant additional (management) complexity, as process can fork, become zombies etc.

**Comment by tsint** [comment]
If off-CPU profiling consumes too many resources, it becomes unsuitable for high-load servers. In many cases, people use profiler to solve problems, and the servers experiencing issues often have insufficient resources. On servers with ample resources, the need for a profiler is not as urgent.

**Comment by florianl** [comment]
> [..]  it becomes unsuitable for high-load servers. In many cases, people use profiler to solve problems, and the servers experiencing issues often have insufficient resources. 

The `Risks` section is exactly about this. And this is the reason, why off-CPU profiling should be added as optional functionality. Meaning, it can be turned on/off as required.
The whole system sampling approach of this agent does work well in production since February 2021 - iirc the date correctly. Also on very high load systems. Consequently one key mitigation strategy to minimize risks for off-CPU profiling is also sampling. In the provided concept scripts, only 3% of the scheduling events (after the dismissal of idle tasks), is considered for further stack unwinding. Depending on the use case and system 3% of scheduling events can be a high or low number. Ultimately the threshold for the sampling (here 3%) should not be hard coded but configurable.

**Review by brancz — APPROVED** [review]
Love this, and we have thought about implementing something like this in a very similar way.

Regarding overhead, I think since this is entirely opt-in, I see no issues regarding it being system-wide, as that's the entire design of the profiler and there's nothing stopping other projects from taking different approaches.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (rockdaboot) [inline design-docs/00001-off-cpu-profiling/README.md:102]:
```suggestion
Attach stack unwinding functionality to the tracepoint `tracepoint:sched:sched_switch`. This
```

File: design-docs/00001-off-cpu-profiling/README.md
Comment (rockdaboot) [inline design-docs/00001-off-cpu-profiling/README.md:194]:
```suggestion
which is crucial for latency analysis.
```

**Review by rockdaboot — APPROVED** [review]

File: design-docs/00001-off-cpu-profiling/README.md
Comment (iogbole) [inline design-docs/00001-off-cpu-profiling/README.md:31]:
```suggestion
and a smaller environmental footprint.
```

**Review by iogbole — COMMENTED** [review]
I'm glad to see this coming to fruition.

**Comment by felixge** [comment]
Looks like we have enough consensus here? Unless there are any more concerns, I (or somebody else) can merge this in the next few days.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:98]: Why not inform userspace? We have safeguards in place to avoid flooding.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:120]: In the actual eBPF implementation, should we also skip kernel-only tasks?

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:143]:
```suggestion
To communicate tasks between the two hooks a `BPF_MAP_TYPE_LRU_HASH` eBPF map should be used with the return of `bpf_get_current_pid_tgid()` as key and the timestamp in nanoseconds as value.
```

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:124]: This should also be configurable in the actual eBPF implementation.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:196]:
```suggestion
Option B might be a bit more complex, as it utilizes one additional hook along with an
```

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:197]:
```suggestion
eBPF map for keeping state across two hooks, compared to Option A with a single hook on
```

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:200]: On supported kernels, we can use trampolines (fentry program) that are faster than kprobes.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:120]: Skipping kernel-only tasks makes sense. I would consider this an implementation detail, which does not have an effect on the overall design/idea of off-cpu profiling.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:124]: With `system_config` there is already a map that can be used to also make this configurable. To showcase the idea and for simplicity, `rand % 100 > 3` was chosen. But I don't see a reason not not make it configurable.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:196]: For the overall idea Option B utilizes two hooks. Only if Option B is compared to Optiona A, then Option B utilizes one  hook more than Option A. But in general, to implement off-cpu profiling functionality, Option B utilizes two additional hooks. Can you clarify the suggestions?

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:200]: This initial design documents tries not to change the minimum supported kernel version and tries to bring the off-cpu functionality to the given circumstances. With newer kernel version things can be done in a simpler way in more places. Using these advanced capabilities for newer kernels should be a future iteration.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:98]: For simplicity userspace should not be informed at this stage. While there are safeguards in place it is hard to predict the impact and name numbers. So maybe informing of userspace can be a future iteration when the design and idea is implemented and such iterations can be backed by benchmarks?

Comment (christos68k) [reply design-docs/00001-off-cpu-profiling/README.md:98]: This should be explored in the initial iteration. Our default rate limiting is fairly conservative and I don't anticipate major issues, but if that's not the case we need to know about it and maybe introduce an additional rate limiting option.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [reply design-docs/00001-off-cpu-profiling/README.md:196]: Updated suggestion to remove "additional"

**Comment by florianl** [comment]
With https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/196 I have opened a draft, that backs this RFC. Once this RFC got merged, I will continue with https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/196.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:120]: I do consider this an implementation detail as it does not have an effect on the overall designs of the proposal. Skipping certain profiles just reduces the work that is done and makes sure the component is not using resources, without providing value for the user.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:98]: The risks section was written in mind with the potential downsides of off-CPU profiling and how they can be mitigated. Overall, the mitigation strategies are just a suggestion and I would also consider them just an implementation detail. Otherwise these mitigation strategies should have been part of `Option A` and/or `Option B`, if they impact the design of the proposals. 

https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/196 just reports every unknown PID to userspace and over the last four weeks I didn't run into issues :tm:. When writing this design doc, I wanted to be open about potential risks and how they can be mitigates. Back then, I had a limiter of reporting unknown PIDs to userspace in place. So far, it seems I wasn't right with the risk.

**Comment by rockdaboot** [comment]
Forgot to mention before: I'm voting for option B

File: design-docs/00001-off-cpu-profiling/README.md
Comment (christos68k) [inline design-docs/00001-off-cpu-profiling/README.md:177]: I am not sure what is meant by aggregation here, e.g. I'm assuming the sampling approach will also aggregate data in maps which will be periodically cleared by userspace.

Then, how can aggregation in the eBPF space that is not sampling (e.g. unwinding every call) be more performant than the sampling approach which will only unwind a subset of these calls?

**Review by christos68k — APPROVED** [review]

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:83]: Naming frequencies might suggest to readers, that the Linux scheduler is called in some constant frequency - which is not correct. Usually, the scheduler provides timeslots for processing to events. The size of the timeslots is configurable, via `/proc/sys/kernel/sched_rr_timeslice_ms`, and is by default 100 ms. Often it is a problem, that events don't make full use of these timeslots - e.g. by calling a syscall or running into a lock.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:98]: Applied the suggestion with some rewording.

File: design-docs/00001-off-cpu-profiling/README.md
Comment (florianl) [reply design-docs/00001-off-cpu-profiling/README.md:177]: > I am not sure what is meant by aggregation here (merging of events?), e.g. I'm assuming the sampling approach will also aggregate data in maps which will be periodically cleared by userspace.

On the eBPF side no aggregation or merging of events for off-CPU data is intended or planed. Once off-CPU hooks are triggered the stack unwinding is using existing eBPF programs and reporting.
In stead of reusing the existing stack unwinding and reporting approach, it would be possible to aggregate off-CPU data and just report merged events to user space. My personal suggestion and the purpose of this document is, not to go this path.

> Then, given that unwinding takes up the majority of eBPF processing time, how can aggregation in the eBPF space that is not sampling (e.g. unwinding every call) be more performant than the sampling approach which will only unwind a subset of these calls?

The overhead of the sampling approach is directly related to the sampling threshold. In the above shown concepts, it is 3% and in https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/196/ it is configurable by the user via the CLI flag `-off-cpu-threshold`. The higher this sampling threshold, the more overhead is introduced by the sampling approach.
To follow the concepts as shown in this document, if 3 is replaced with 99, then nerarly every scheduling event triggers the off-CPU hooks. In such a case it would make sense, to aggregate off-CPU data in eBPF space for a summarized report to user space, instead of reporting every single stack unwinding.

Overall, sampling and aggregation can be done in various places and it depends on configuration and workload of the system. This paragraph should make the user aware of alternative concepts and focuses on the eBPF space for the sampling vs. aggregation approach. The proposed concepts should be fit most use cases - optimization for special cases is always an option.

**Review by fabled — APPROVED** [review]
LGTM, and I am voting also option B. Seems all agree on this, so we have consensus.

@florianl Could you update the Decision in the end, and potential Meta in the beginning too. We can merge this right after this. Thank you!

**Comment by florianl** [comment]
Added the decision to go forward with Option B.

**Pull request merged**
